### PR TITLE
Show (optional) description for each macro in the help formaction

### PR DESCRIPTION
**Pull request by dennisschagt**
Fixes https://github.com/newsboat/newsboat/issues/228.

This is done as a small preparation for https://github.com/newsboat/newsboat/issues/1165 (idea: https://github.com/newsboat/newsboat/issues/1165#issuecomment-696674754).

Example output:
image
Based on config:
```
macro x set browser "mpv" ; open-in-browser -mark-read      -- "Open with mpv"
macro y set browser "mpv" ; open-in-browser-and-mark-read   -- "Open with mpv and mark read"
macro z set feedlist-title-format "changed title"
```

TODO:
- [x] Update docs

### CODE REVIEW

**Review by der-lyse — COMMENTED** [review]
This is brilliant! I tried it and it works perfectly.

It seems that, double quotes cannot be used in the comment. Escaping them with a backslash just strips everything after the backslash. The config:

    macro x quit -- "foo \" bar\"

Will be rendered as:

    <prefix-key>x                           foo \

Not sure if this intended or not. Personally, I could easily live with this.

Maybe it would be useful, to replace the `<prefix key>` by the real configured or default prefix key. So that e.g. it would be shown as:

    ,x                                      foo \

I reckon we could also remove a lot of the whitespace between the macro key binding and its description.

**Review by Minoru — CHANGES_REQUESTED** [review]
I'm happy how this looks UI-wise, but I got a few suggestions regarding the implementation.

In addition to the problem that @der-lyse spotted, there is one more: it's not an error to put `--` in the middle of the config line, e.g.:

```
macro a set browser firefox -- https://DownForEveryonneOrJustMe.com/%u ; open-in-browser ; set browser firefox
```

Since we don't display the list of operations next to a macro, it'd be hard for the user to realize that this macro only runs `set browser firefox`. I think this line should result in an error, so the user can go and fix it right away.

To me, it looks like both problems stem from the same source: the `-- desc` thing is not part of the `operation_sequence` parser. I think that parser should be generalized to parse *everything* that comes after `macro x`. But I won't be too surprised if you come up with an even better solution :)

Also, this PR contains a new translatable string, "Macros", and we're less than two weeks from the 2.23 release, so even if this is ready within a few days, I'll hold off merging until after 2.23 — that gives translators an opportunity to get to 100% translated.

File: rust/libnewsboat/src/keymap.rs
Comment (dennisschagt) [reply rust/libnewsboat/src/keymap.rs]: Fixed in https://github.com/newsboat/newsboat/pull/1516/commits/6dd1acb5233339a83ccd02379e85114a034cd4d9

**Comment by dennisschagt** [comment]
Good points, thanks for the reviews!

I rebased onto current master branch but kept the existing 5 commits unchanged (only commit hashes got changed due to the rebase).

> It seems that, double quotes cannot be used in the comment. Escaping them with a backslash just strips everything after the backslash.

Fixed in https://github.com/newsboat/newsboat/pull/1516/commits/b994d4db350bb0dd8286a3e8c5ee2c9a41164379.

> Maybe it would be useful, to replace the <prefix key> by the real configured or default prefix key.

Sounds good to me, however, what should we display if the `macro-prefix` command is bound to multiple keys? (or not bound at all)

> I reckon we could also remove a lot of the whitespace between the macro key binding and its description.

I added the whitespace to line out the descriptions with the descriptions of the regular key bindings.
However, I'm fine reducing the whitespace if you, or anyone else, dislikes the amount of whitespace.

> I think that parser should be generalized to parse everything that comes after macro x. 

Done in https://github.com/newsboat/newsboat/pull/1516/commits/bfa3fb2be9b2d57b09bbcd5b4f2a72571cf5ab51.

> Since we don't display the list of operations next to a macro, it'd be hard for the user to realize that this macro only runs set browser firefox. I think this line should result in an error, so the user can go and fix it right away.

Done.
However, the error does not explain the exact problem, current output:
```
Loading configuration...Error while processing command `macro a set browser firefox -- https://DownForEveryonneOrJustMe.com/%u ; open-in-browser ; set browser firefox' (/home/dennis/projects/newsboat/densch-envs/default/.newsboat/config line 85): too few parameters.
```

Is that error acceptable?

File: src/keymap.cpp
Comment (Minoru) [inline src/keymap.cpp:799]: That's a nice one :) This shouldn't have compiled — this is [a designated initializer](https://en.cppreference.com/w/cpp/language/aggregate_initialization), which is only available in C++20 (and C99). I suppose it works because we don't use `-pedantic`, and since this feature is available in C99, even GCC 5.0 supports it.

No need to change this, just wanted to make this note so you're aware.

**Review by Minoru — CHANGES_REQUESTED** [review]
This is good progress, thanks for working on this! However, there are still a few loose ends to tie off.

**Comment by Minoru** [comment]
> what should we display if the macro-prefix command is bound to multiple keys? (or not bound at all)

That's a good point. I think we should always say `<macro-prefix>`. Since the user is looking at the list of bindings, they can easily scan up the page and find what `macro-prefix` is bound to.

In fact, perhaps we shouldn't even translate `macro-prefix`, to make this connection more obvious.

> I reckon we could also remove a lot of the whitespace between the macro key binding and its description.

I vote for less whitespace, too. Maybe two spaces between the key and the description? Or maybe one tab.

> the error does not explain the exact problem, current output:
>
>    Loading configuration...Error while processing command `macro a set browser firefox -- https://DownForEveryonneOrJustMe.com/%u ; open-in-browser ; set browser firefox' (/home/dennis/projects/newsboat/densch-envs/default/.newsboat/config line 85): too few parameters.
>
> Is that error acceptable?

Not very :) I think it's not too hard to improve though. There are two error cases:

1. the list of ops is empty — that's what `TOO_FEW_PARAMS` designates. We can just keep doing this;
2. `operation_description` parser failed. This can be caused either by a double-dash in the middle of a list, or e.g. by a missing double quote at the end of the description. I don't see an easy way to distinguish between those two cases, and I think we can lump them together. Just tell the user that we couldn't parse the macro description, and refer them to the docs on `macro` for details. Even if the user put double-dash into their macro by mistake, they'll discover that it has meaning to Newsboat, and will learn that they can write `"--"` to avoid this error. And if they *did* intend to write a macro description, they'll be reminded of all the minutae like the double quotes.

    We'll need a new constant to designate this case.

Since we aren't passing any error messages across the FFI, we don't need additional structs like with `utils::read_text_file` — the second case can be signalled by a `bool` out-parameter, same as the description is passed.

Does that sound reasonable?

File: src/keymap.cpp
Comment (dennisschagt) [reply src/keymap.cpp:799]: I wasn't aware, thanks!

**Comment by dennisschagt** [comment]
@Minoru: I believe I handled all of your inline review comments.

Still TODO:
- [ ] Update documentation of `macro`
- [ ] Print a more appropriate error than "too few parameters"

Regarding the error handling/message:
> Does that sound reasonable?

Sounds good, I plan to implement that soon.

**Review by Minoru — COMMENTED** [review]
Yup, you've handled them all. Thanks!

What's the rationale for 087e9458c1bcd6afe8226e5208bfe9a726477437? It doesn't seem to be connected to any discussion in this PR, and I'm a bit suspicious since it removes a test :)

File: rust/libnewsboat/src/keymap.rs
Comment (dennisschagt) [reply rust/libnewsboat/src/keymap.rs]: Nice idea, done: https://github.com/newsboat/newsboat/pull/1516/commits/3a1743c201ef960e07c614414cd3c11c94801633

File: rust/libnewsboat/src/keymap.rs
Comment (dennisschagt) [reply rust/libnewsboat/src/keymap.rs:120]: No problem, didn't take that much effort

**Comment by dennisschagt** [comment]
> What's the rationale for 087e945? It doesn't seem to be connected to any discussion in this PR, and I'm a bit suspicious since it removes a test :)

I found an issue when an argument started with `--"`. That would be consumed by the `unquoted_token` parser, even if it was part of a description, e.g. `--"sample description"`. (It only got consumed if there was whitespace in front of the dashes. Now tested by https://github.com/newsboat/newsboat/commit/6afc2861326cb9758b0666c719d15f2aa372fef5#diff-e1f94a504c2bba8618d679729ef8ca8b7217cf47a65cc89bed682b08b56e4f34R401-R404).
As it doesn't make sense to have quotes inside of an unquoted token, I think it makes sense to disallow them altogether.

**Review by Minoru — APPROVED** [review]
Okay then, I think this PR is can be merged! Since there are fixup commits, I'll leave it up to you @dennisschagt to squash them before you merge it.

Thank you very, very much for all the work you've put into this pull request! I really appreciate the attention to detail and the commitment to make the design as flawless as possible.

**Comment by dennisschagt** [comment]
Thanks!

Just rebased and enabled auto-merge.

**Pull request merged**
